internal/pkg/action/operator_list_custom_resources.go

```diff
@@ -0,0 +1,113 @@
+	"strings"
+
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/api/pkg/operators/v2alpha1"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type CustomResourceLister interface {
+	List(ctx context.Context, ownedCRDS []v1alpha1.CRDDescription) (metav1.PartialObjectMetadataList, error)
+}
+
+type CustomResourceDefinitionFinder interface {
+	FindOperator(ctx context.Context, packageName, namespace string) (*v2alpha1.Operator, error)
+	Unzip(ctx context.Context, operator *v2alpha1.Operator) ([]v1alpha1.CRDDescription, error)
+}
+
+// OperatorListCustomResources knows how to find and list custom resources given a package name and namespace.
+type OperatorListCustomResources struct {
+	config  *Configuration
+	ListAll bool
+	PackageName string
+	Namespace string
+}
+
+func NewOperatorListCustomResources(cfg *Configuration) *OperatorListCustomResources {
+	return &OperatorListCustomResources{
+		config: cfg,
+		Namespace: cfg.Namespace,
+	}
+}
+
+// FindOperator finds an operator object on-cluster provided a package and namespace. Returns a copy of the operator.
+func (o *OperatorListCustomResources) FindOperator(ctx context.Context) (v2alpha1.Operator, error) {
+	opKey := types.NamespacedName{
+		Name:      o.PackageName,
+		Namespace: o.Namespace,
+	}
+	operator := v2alpha1.Operator{}
+	err := o.config.Client.Get(ctx, opKey, &operator)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return v2alpha1.Operator{}, fmt.Errorf("could not find operator for package %s in namespace %s", o.PackageName, o.Namespace)
+		}
+		return v2alpha1.Operator{}, err
+	}
+	return operator, nil
+}
+
+// Unzip finds the CSV referenced by the provided operator and then inspects the spec.customresourcedefinitions.owned
+// section of the CSV to return a list of APIs that are owned by the CSV.
+func (o *OperatorListCustomResources) Unzip(ctx context.Context, operator *v2alpha1.Operator) ([]v1alpha1.CRDDescription, error) {
+	csv := v1alpha1.ClusterServiceVersion{}
+	csvKey := types.NamespacedName{}
+
+	if operator.Status.Components == nil {
+		return nil, fmt.Errorf("could find underlying components for operator %s", operator.Name)
+	}
+	for _, resource := range operator.Status.Components.Refs {
+		if resource.Kind == v1alpha1.ClusterServiceVersionKind {
+			csvKey.Name = resource.Name
+			csvKey.Namespace = resource.Namespace
+			break
+		}
+		return nil, fmt.Errorf("could find underlying CSV for operator %s", operator.Name)
+	}
+
+	err := o.config.Client.Get(ctx, csvKey, &csv)
+	if err != nil {
+		return nil, fmt.Errorf("could not get %s CSV on cluster: %s", csvKey.String(), err)
+	}
+
+	// check if owned CRDs are defined on the csv first
+	if reflect.DeepEqual(csv.Spec.CustomResourceDefinitions, v1alpha1.CustomResourceDefinitions{}) || len(csv.Spec.CustomResourceDefinitions.Owned) == 0 {
```

Comment:
Looks like `csv.Spec.CustomResourceDefinitions` is a concrete struct value, so probably no need to check that its empty since it can never be `nil`. Therefore, you can directly check `len(csv.Spec.CustomResourceDefinitions.Owned) == 0` without the first check.